[feature-wip](MTMV) Support importing data to materialized view with multiple tables

chenlinzhong opened the pull request:
# Proposed changes

Issue Number: close #xxx

## Problem summary

impl import data to multi table materialized view table,mainly include four steps
1、create tmp table
2、insert data to tmp table though insert into select stmt
3、swap tmp table with origin table
4、update task info and wrtie log

```
create table t_user(
     event_day DATE,
     id bigint,
     username varchar(20)
)
DISTRIBUTED BY HASH(id) BUCKETS 10 
PROPERTIES (
   "replication_num" = "1"
 );
insert into  t_user values("2022-10-26",1,"clz");
insert into  t_user values("2022-10-28",2,"zhangsang");
insert into  t_user values("2022-10-29",3,"lisi");
create table t_user_pv(
    event_day DATE,
    id bigint,
    pv bigint
)
DISTRIBUTED BY HASH(id) BUCKETS 10 
PROPERTIES (
   "replication_num" = "1"
 );
insert into  t_user_pv  values("2022-10-26",1,200);
insert into  t_user_pv  values("2022-10-28",2,200);
insert into  t_user_pv  values("2022-10-28",3,300);
```
### create multi view table 
```
DROP MATERIALIZED VIEW  if exists multi_mv;
CREATE MATERIALIZED VIEW  multi_mv
BUILD IMMEDIATE 
REFRESH COMPLETE 
start with "2022-10-27 19:35:00"
next  60 second
KEY(username)   
DISTRIBUTED BY HASH (username)  buckets 1
PROPERTIES ('replication_num' = '1') 
AS 
select t_user.username, t_user_pv.pv  from t_user, t_user_pv where t_user.id=t_user_pv.id;
```
## show mtmv task \G
<img width="892" alt="image" src="https://user-images.githubusercontent.com/11487604/202708285-baaa082e-62ef-4ff4-9a34-04a3f36e16be.png">
<img width="360" alt="image" src="https://user-images.githubusercontent.com/11487604/202708355-f0df8e70-df86-462e-ab37-cb6d91b35951.png">

## search run log
grep --color ${taskid} fe.log
![image](https://user-images.githubusercontent.com/11487604/202710580-057e74e5-1b85-4879-beff-aa1f8d73d9e7.png)




## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

chenlinzhong committed:
impl import data to multi table materialized view table,mainly include four steps

1、create tmp table

2、insert data to tmp table though insert into select stmt

3、swap tmp table with origin table

4、update task info and wrtie log

adonis0147 commented:
This is a sub task of #7503

chenlinzhong committed:
impl import data to multi table materialized view table,mainly include four steps

1、create tmp table

2、insert data to tmp table though insert into select stmt

3、swap tmp table with origin table

4、update task info and wrtie log

adonis0147 commented on fe/fe-core/src/main/java/org/apache/doris/catalog/DatabaseIf.java:
```diff
@@ -120,6 +120,16 @@ default T getTableOrMetaException(String tableName, TableIf.TableType tableType)
         return table;
     }
 
+    default T getTableViewOrMetaException(String tableName) throws MetaNotFoundException {
+        T table = getTableOrMetaException(tableName);
+        if (table.getType() != TableIf.TableType.OLAP && table.getType() != TableIf.TableType.MATERIALIZED_VIEW) {
+            throw new MetaNotFoundException(
+                    "table type is not olap or materialized view, tableName=" + tableName + ", type="
+                            + table.getType());
+        }
+        return table;
+    }
+
```
I think the following interface is better due to it is more generic.
```suggestion
    default T getTableOrMetaException(String tableName, List<TableIf.TableType> tableTypes)
            throws MetaNotFoundException {
        T table = getTableOrMetaException(tableName);
        if (!tableTypes.contains(table.getType())) {
            throw new MetaNotFoundException(
                    "Tye type of " + tableName + " doesn't match, expected data tables=" + tableTypes);
        }
        return table;
    }
```

adonis0147 commented on fe/fe-core/src/main/java/org/apache/doris/common/FeConstants.java:
```diff
@@ -79,4 +79,5 @@ public class FeConstants {
     public static String FS_PREFIX_HDFS = "hdfs";
     public static String FS_PREFIX_FILE = "file";
     public static final String INTERNAL_DB_NAME = "__internal_schema";
+    public static String TEMP_MATERIZLIZE_DVIEW_PREFIX = "doris_internal_tmp_multi_view_";
```
I don't think the name `multi view` is suitable, because we just create a one temp view for each job.

adonis0147 commented on fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVJobFactory.java:
```diff
@@ -23,21 +23,31 @@
         BuildMode buildMode = materializedView.getBuildMode();
-        MVRefreshTriggerInfo triggerInfo =  materializedView.getRefreshInfo().getTriggerInfo();
+        MVRefreshTriggerInfo triggerInfo = materializedView.getRefreshInfo().getTriggerInfo();
+        //cannnot generate job when create temp view table
```
```suggestion
        //can not generate a job when creating a temp materialized view.
```

adonis0147 commented on fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVTaskExecutorPool.java:
```diff
@@ -68,7 +68,9 @@ public void executeTask(MTMVTaskExecutor taskExecutor) {
                 task.setState(TaskState.FAILED);
                 task.setErrorCode(-1);
-                task.setMessage(lastExceptionString);
+                if (lastExceptionString.length() > 0) {
```
Does this branch necessary?

adonis0147 commented on fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVTaskProcessor.java:
```diff
@@ -17,13 +17,249 @@
-        LOG.info("run mv logic here.");
+        String taskId = context.getTask().getTaskId();
+        long jobId = context.getJob().getId();
+        LOG.info("run mv logic start, task_id:{}, jobid:{}", taskId, jobId);
```
I don't think the log message `run mv logic start` is suitable.

adonis0147 commented on fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVTaskProcessor.java:
```diff
@@ -17,13 +17,249 @@
+                .getDbOrAnalysisException(context.getTask().getDbName());
+        MaterializedView table = (MaterializedView) db.getTableOrAnalysisException(tableName);
+        if (!table.tryMvTaskLock()) {
+            LOG.warn("run mv task  failed, taskid:{}, jobid:{}, msg:{}", taskId, jobId, "get lock fail");
```
```suggestion
            LOG.warn("run MTMV task failed, taskid:{}, jobid:{}, msg:{}", taskId, jobId, "get lock fail");
```

adonis0147 commented on fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVTaskProcessor.java:
```diff
@@ -17,13 +17,249 @@
+        } catch (IOException e) {
+            LOG.warn("execSQL failed, taskid:{}, msg:{}, stmt:{}", context.getTask().getTaskId(), e.getMessage(),
+                    originStmt);
+        } catch (UserException e) {
+            LOG.warn("execSQL failed, taskid:{}, msg:{}, stmt:{}", context.getTask().getTaskId(), e.getMessage(),
+                    originStmt);
+        } catch (Throwable e) {
+            LOG.warn("execSQL failed, taskid:{}, msg:{}, stmt:{}", context.getTask().getTaskId(), e.getMessage(),
+                    originStmt);
```
We can merge these exception handlings by using `catch (Throwable e)`

chenlinzhong committed:
impl import data to multi table materialized view table,mainly include four steps

1、create tmp table

2、insert data to tmp table though insert into select stmt

3、swap tmp table with origin table

4、update task info and wrtie log

chenlinzhong committed:
impl import data to multi table materialized view table,mainly include four steps

1、create tmp table

2、insert data to tmp table though insert into select stmt

3、swap tmp table with origin table

4、update task info and wrtie log

zhouhoo commented:
Hi, could  anyone tell when the great feature of multi table materialized view can realsed?  Is it possible to see it in doris next version?

chenlinzhong commented:
> Hi, could anyone tell when the great feature of multi table materialized view can realsed? Is it possible to see it in doris next version?

there is no plan to release at present, but it will be as soon as possible, he development can be completed next month

adonis0147 reviewed (approved):
LGTM
